kong/tools/wrpc.lua

```diff
@@ -0,0 +1,685 @@
+    self.smph:post()
+  end
+
+  local function expire(self)
+    self:handle_error("timeout", "timeout")
+  end
+
+  function semaphore_waiter()
+    return {
+      smph = semaphore.new(),
+      deadline = ngx.now() + DEFAULT_EXPIRATION_DELAY,
+      handle = handle,
+      handle_error = handle_error,
+      expire = expire,
+    }
+  end
+end
+
+
+local remote_waiter
+do
+  local function handle(self, payload)
+    channel.put_back(self.dict, self.name, pb.encode("wrpc.PayloadV1", payload))
+  end
+
+  local function handle_error(self, etype, errdesc)
+    channel.put_back(self.dict, self.name, pb.encode("wrpc.PayloadV1", {
+      mtype = "MESSAGE_TYPE_ERROR",
+      error = {
+        etype = etype,
+        description = errdesc,
+      }
+    }))
+  end
+
+  function remote_waiter(dict, name)
+    return {
+      dict = dict,
+      name = name,
+      deadline = ngx.now() + DEFAULT_EXPIRATION_DELAY,
+      handle = handle,
+      handle_error = handle_error,
+      raw = true,
+    }
+  end
+end
+
+
+local function merge(a, b)
+  if type(b) == "table" then
+    for k, v in pairs(b) do
+      a[k] = v
+    end
+  end
+
+  return a
+end
+
+local function proto_searchpath(name)
+  return package.searchpath(name, "kong/include/?.proto;/usr/include/?.proto")
+end
+
+--- definitions for the transport protocol
+local wrpc_proto
+
+
+local wrpc_service = {}
+wrpc_service.__index = wrpc_service
+
+
+--- a `service` object holds a set of methods defined
+--- in .proto files
+function wrpc.new_service()
+  if not wrpc_proto then
+    local wrpc_protofname = assert(proto_searchpath("wrpc.wrpc"))
+    wrpc_proto = assert(grpc.each_method(wrpc_protofname))
+  end
+
+  return setmetatable({
+    methods = {},
+  }, wrpc_service)
+end
+
+--- Loads the methods from a .proto file.
+--- There can be more than one file, and any number of
+--- service definitions.
+function wrpc_service:add(service_name)
+  local annotations = {
+    service = {},
+    rpc = {},
+  }
+  local service_fname = assert(proto_searchpath(service_name))
+  local proto_f = assert(io.open(service_fname))
+  local scope_name = ""
+
+  for line in proto_f:lines() do
+    local annotation = line:match("//%s*%+wrpc:%s*(.-)%s*$")
+    if annotation then
+      local nextline = proto_f:read("*l")
+      local keyword, identifier = nextline:match("^%s*(%a+)%s+(%w+)")
+      if keyword and identifier then
+
+        if keyword == "service" then
+          scope_name = identifier;
+
+        elseif keyword == "rpc" then
+          identifier = scope_name .. "." .. identifier
+        end
+
+        local type_annotations = annotations[keyword]
+        if type_annotations then
+          local tag_key, tag_value = annotation:match("^%s*(%S-)=(%S+)%s*$")
+          if tag_key and tag_value then
+            tag_value = tag_value
+            local tags = type_annotations[identifier] or {}
+            type_annotations[identifier] = tags
+            tags[tag_key] = tag_value
```

Comment:
tags seem not used

Reply:
they're part of the `annotation` structure, it's put into `type_annotations`, which is a subtable of `annotations`.  after scanning the file, it's read to get the `service_id` and `rpc_id` on the `.each_method()` loop